Validate generated expressions of partitioned by columns

BaurzhanSakhariev opened the pull request:
Validate generated expressions of partitioned by columns.

Also, do it before creating a partition so that invalid rows don't cause abandoned partitions.

Closes https://github.com/crate/crate/issues/14304

Supersedes https://github.com/crate/crate/pull/14317

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dsl/projection/ColumnIndexWriterProjection.java:
```diff
@@ -229,6 +248,12 @@ public int hashCode() {
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
 
+        if (out.getVersion().onOrAfter(Version.fromString("5.3.4"))) {
+            out.writeVInt(partitionedByColumns.size());
```
Maybe `.after(Version.5_3_3)` would be better in order not to use constant instead of `String`?

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/ColumnIndexWriterProjector.java:
```diff
@@ -76,6 +79,7 @@ public ColumnIndexWriterProjector(ClusterService clusterService,
                                       @Nullable Symbol routingSymbol,
                                       ColumnIdent clusteredByColumn,
                                       List<Reference> columnReferences,
+                                      @NotNull List<Reference> partitionedByColumns,
                                       List<Input<?>> insertInputs,
```
this comes from `DocTableInfo.partitionedByColumns` which afaik is not null, also all streaming and all handling components receiving it doesn't do null checks.

Entry point is `DocIndexMetadata.getPartitionedBy`, looks like safe to say it's not null?

BaurzhanSakhariev commented on docs/appendices/release-notes/unreleased.rst:
```diff
@@ -293,3 +293,9 @@ Fixes
     +----+-------------+
     |  1 | {"x": null} |
     +----+-------------+
+
+- Fixed an issue which caused ``INSERT INTO`` statements to skip generated
+  expression validation for partitioned columns.
+
+- Fixed an issue which caused ``INSERT INTO`` statements to leave behind
+  empty partitions if NULL or CHECK constraint on partitioned by column failed.
```
Not sure whether we need 2 entries but it looks to me like a separate issues:

1. Validation of gen expression for partitioned by columns was not done at all. https://github.com/crate/crate/issues/14304. we cannot say it left some partitions behind

2. not null check was done on 5.3 but partitions were left behind

Fix tackles both because it about 
do all validation and do it early.

"all" covers gen expressions, others checks are done
"early" covers abandoned partitions

BaurzhanSakhariev commented:
Hi @seut, could you please take a look?

As discussed, doing only partitioned by column validation before creating a partition.

Roughly the idea is:
1. Use helper method `Indexer.createConstraintCheck` similar to how it's done in `InsertFromValues`. Tweak it to accept targetColumns and their values as for insert-from-subquery we pass it explicitly don't derive from `indexItem`.

2. `InsertFromValues` had concrete control over rows and we used to create validatorCache. In insert-from-subquery I was not sure how to clean such cache later and decided to put that cache to the `ShardedRequests` and clean it on close. All items kinda belongs to `ShardedRequests` as they fill it up and such I thought validation cache entries are relevant for it.

3.  Indexer.createConstraintCheck needs tableInfo, had to pass it through multiple components, starting from ProjectionToProjector visitor, many changes in this PR is about that propagation. Let me know if there is an easier way to get table info in grouper.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -1869,4 +1869,53 @@ public void test_generated_sub_column_provided_null_is_validated() {
+        execute("insert into t (a, b, c, d) select null, 12, 1, 1");
+        assertThat(response.rowCount()).isEqualTo(0L);
+
+
+        // We need to ensure that check is done before partition creation.
+        // If check is done too late, INSERT statement might work as expected and reject invalid records but invalid partitions will left behind.
+        // At this point all failing scenarios are done and haven't written anything.
+        // Checking that neither of them has created a partition.
+        Metadata updatedMetadata = cluster().clusterService().state().metadata();
+        String tableTemplateName = PartitionName.templateName("doc", "t");
+        for (ObjectCursor<String> cursor : updatedMetadata.indices().keys()) {
+            String indexName = cursor.value;
+            assertThat(PartitionName.templateName(indexName)).isNotEqualTo(tableTemplateName);
```
as mentioned in https://github.com/crate/crate/pull/14373#issuecomment-1625363870,

row can fail because of regular columns (and parted columns can have valid values). In this case we still shouldn't create partitions but such case covered already in `test_insert_from_subquery_clustered_by_null_rejected`

BaurzhanSakhariev commented:
Hi @seut could you please take a look?

As discussed, I tried implementing only partitioned by column validation before creating a partition in https://github.com/crate/crate/pull/14373/commits/5a72c8bada09257f2b590b43764476f78e341132. 

Then run into 2 issues:
1. Partitioned column can refer to regular column, so need to set targetColumns not only partitionedByColumns but allTargetColumns. This was minor, as I could send **all** values but introduce some flag and check **only partitioned**.
2. Major issue was potential row failure because of regular columns. See https://github.com/crate/crate/pull/14373#issuecomment-1625363870 for details.

I implemented then early validation of all columns in https://github.com/crate/crate/pull/14373/commits/536d72409d41857a4728294d2173bc3d3b72594f.

Difference with https://github.com/crate/crate/pull/14317 I that this time I haven't removed validation from the indexing phase in order not to break UPDATE and nested column checks.


Roughly the idea is:

1. Use helper method `Indexer.createConstraintCheck` similar to how it's done in `InsertFromValues`. Tweak it to accept targetColumns.

2. `InsertFromValues` has concrete control over rows and we used to create validatorCache. In insert-from-subquery I was not sure how to clean such cache later and decided to put that cache to the `ShardedRequests` and clean it on `close`. All items kinda belongs to `ShardedRequests` as they fill it up and such I thought validation cache entries are relevant for it.

3.`Indexer.createConstraintCheck` needs `tableInfo`, had to pass it through multiple components, starting from ProjectionToProjector visitor, many changes in this PR is about that propagation. Let me know if there is an easier way to get table info in grouper.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java:
```diff
@@ -162,7 +179,24 @@ public TItem apply(ShardedRequests<TReq, TItem> shardedRequests, Row row, Boolea
             RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
             ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            IndexItem indexItem = (IndexItem) item;
             if (shardLocation == null) {
+                if (allTargetColumns != null) {
+                    // Validation is done before creating an index in order to ensure
```
We have to do validation of all columns and it means we might do it twice, second time on indexing

 However it's true only for tables with partitioned by columns. I was thinking about some optimization for the check to not run if `table.partitionedBy.empty == false` but I think  `shardLocation == null `check already ensures that we do "double" validation only for partitioned tables.

seut replied:
Can't we pass in a `BiConsumer<String, IndexItem>` instead to all components instead? Looks to me that this is possible (but maybe I miss something) and would avoid that all the required arguments needed by `checkContraints` are passed through the whole chain.

BaurzhanSakhariev replied:
We cannot have a single `BiConsumer` for all components since `InsertFromValues` has it's own values for targetColumns, cache, tableInfo and contexts.

Thus, created dedicated BiConsuer-s for insert-from-subquery and insert-from-values in https://github.com/crate/crate/pull/14373/commits/b2c2d4478779613feaf53f2bbfb84ddd450f53d4.

Got rid from references and tableInfo, thanks!

seut commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -254,24 +257,28 @@ public void execute(DependencyCarrier dependencies,
 
             try {
                 // Primary Key and CLUSTERED BY check is already done in grouper -> RowShardResolver, both cannot be null.
+                // checkConstraintsOnGeneratedSource is also done in grouper.apply but only for partitioned tables.
+                // It's important to keep check below so that
+                // insert from values into regular tables doesn't leave partitions behind in case of validation failure.
```
The comment is confusing as on regular tables (non-partitioned), no bad partition exists without proper validation.
I think the validation here is done for optimization reasons as all values already exists and thus we avoid sending invalid ones to the shards?

BaurzhanSakhariev replied:
thanks, comment was totally wrong. 
What I had in mind was keeping old execution logic for Insert from values into regular tables.

Indeed, abandoned partitions issue is not related to this case.

I think we can even drop this check now, wdyt? 

For now I kept it and updated the comment that we just want to fail early and don't send invalid values to the shards.

seut commented on server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java:
```diff
@@ -55,11 +58,13 @@ public final class ShardedRequests<TReq extends ShardRequest<TReq, TItem>, TItem
     public ShardedRequests(Function<ShardId, TReq> requestFactory, RamAccounting ramAccounting) {
         this.requestFactory = requestFactory;
         this.ramAccounting = ramAccounting;
+        this.validatorsCache = new HashMap<>();
```
Could be done on variable declaration.

BaurzhanSakhariev replied:
Done but in different place, since I moved cache creation/clearance to another place. 
See https://github.com/crate/crate/pull/14373#discussion_r1260651275

seut commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -486,11 +498,13 @@ public List<CompletableFuture<Long>> executeBulk(DependencyCarrier dependencies,
-                                                   DocTableInfo tableInfo,
-                                                   PlannerContext plannerContext,
-                                                   HashMap<String, Consumer<IndexItem>> validatorsCache) throws Throwable {
+    public static void checkConstraintsOnGeneratedSource(@Nullable IndexItem indexItem,
```
Not dedicated to this change, but the existing name is a bit missleading nowadays.
```suggestion
    public static void checkConstraints(@Nullable IndexItem indexItem,
```

seut reviewed (commented):
looks good in general

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/IndexWriterProjector.java:
```diff
@@ -65,11 +66,13 @@
 import io.crate.metadata.Reference;
 import io.crate.metadata.TransactionContext;
 
+@Deprecated
```
I wanted to highlight that we can just pass null and skip this early validation for COPY FROM as in another PR `IndexWriterProjector` is being removed together with _raw optimization removal. 

I removed `@deprecated` as right now there is no alternative.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dsl/projection/ColumnIndexWriterProjection.java:
```diff
@@ -44,11 +44,10 @@
     private final boolean ignoreDuplicateKeys;
     private final Map<Reference, Symbol> onDuplicateKeyAssignments;
     private final List<Reference> allTargetColumns;
+    private final List<Symbol> targetColsSymbols;
```
One is References and another is InputColumns with values.

Similar to the previous pair
```
private final List<Symbol> targetColsSymbolsExclPartition;
private final List<Reference> targetColsExclPartitionCols;
```

BaurzhanSakhariev replied:
got rid of symbols/input columns by deriving them from References, see comment https://github.com/crate/crate/pull/14373#discussion_r1262443382

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java:
```diff
@@ -55,11 +58,13 @@ public final class ShardedRequests<TReq extends ShardRequest<TReq, TItem>, TItem
+        this.validatorsCache = new HashMap<>();
+    }
+
+    public HashMap<String, Consumer<IndexItem>> validatorsCache() {
```
> Could this be moved to where it's used, or move the functionality that uses it into this class?

I moved cache to the places which use it in https://github.com/crate/crate/pull/14373/commits/6ea2e7b62296b340adf5304e6815d149620bb3d7

`InsertFromValues` now almost same as master but sharing cache with the grouper which now also needs the cache.
With insert from subquery unfortunately there is a longer chain: it's a grouper using it but flow control happens in ShardUpsertExecutor. so moved cache declaration/clearance there.

BaurzhanSakhariev commented:
I updated `Version` check in https://github.com/crate/crate/pull/14373/commits/7360012b7d0442640a3c12c5a03c24a37146c651 to reflect 5.3 label replacement by 5.4

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dsl/projection/ColumnIndexWriterProjection.java:
```diff
@@ -101,18 +98,18 @@ public ColumnIndexWriterProjection(RelationName relationName,
-            targetColsExclPartitionCols = new ArrayList<>(length);
-            for (int i = 0; i < length; i++) {
-                targetColsExclPartitionCols.add(Reference.fromStream(in));
+        if (in.getVersion().onOrBefore(Version.fromString("5.3.4"))) {
```
Thanks, updated write in a similar way as well in https://github.com/crate/crate/pull/14373/commits/3491ebf5f1cf2ee60e349187831a72063b4babd8

Sorry, my master got stale and `Version.V_5_5_0` hadn't compiled, had to rebase.

seut commented on server/src/test/java/io/crate/planner/InsertPlannerTest.java:
```diff
@@ -229,9 +230,13 @@ public void testInsertFromSubQueryGlobalAggregate() {
-        assertThat(projection.columnSymbols().size(), is(2));
-        assertThat(((InputColumn) projection.columnSymbols().get(0)).index(), is(0));
-        assertThat(((InputColumn) projection.columnSymbols().get(1)).index(), is(1));
+        List<Symbol> columnSymbols = InputColumns.create(
+            projection.allTargetColumns(),
+            new InputColumns.SourceSymbols(projection.allTargetColumns()));
+
+        assertThat(columnSymbols.size(), is(2));
+        assertThat(((InputColumn) columnSymbols.get(0)).index(), is(0));
+        assertThat(((InputColumn) columnSymbols.get(1)).index(), is(1));
```
Please remove, this tests the logic of `InputColumns.SourceSymbols` and thus should not be part of this test.

seut commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -504,29 +519,6 @@ public List<CompletableFuture<Long>> executeBulk(DependencyCarrier dependencies,
         );
     }
 
-    public static void checkConstraints(@Nullable IndexItem indexItem,
```
Can we keep this method please to avoid duplicated code?

seut commented on server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java:
```diff
@@ -105,11 +102,11 @@ public class ShardingUpsertExecutor
     ShardingUpsertExecutor(ClusterService clusterService,
-                           DocTableInfo tableInfo,
-                           @Nullable List<Reference> allTargetColumns,
+                           @Nullable BiConsumer<String, IndexItem> constraintsChecker,
+                           @Nullable Runnable validatorsCacheCleaner,
```
We could just pass in dummy a runnable `() -> {}` and consumer, so any `NULL` check can be avoided,

seut commented on server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java:
```diff
@@ -68,7 +69,10 @@ public final class GroupRowsByShard<TReq extends ShardRequest<TReq, TItem>, TIte
+    private final BiConsumer<String, IndexItem> constraintsChecker;
+
     public GroupRowsByShard(ClusterService clusterService,
+                            @Nullable BiConsumer<String, IndexItem> constraintsChecker,
```
same here, I think passing in a dummy consumer instead of a null would avoid any mistake (null check)

seut commented on server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java:
```diff
@@ -162,7 +169,16 @@ public TItem apply(ShardedRequests<TReq, TItem> shardedRequests, Row row, Boolea
 
             RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
             ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            IndexItem indexItem = (IndexItem) item;
```
The `item` can also be a `ShardDeleteRequestItem` which doesn't implement `IndexItem` so this would fail for delete requests. Such we must surround the checker call with `if (item instanceof IndexItem indexItem) {}`

BaurzhanSakhariev replied:
Grouper is used only for inserts/`ShardingUpsertExecutor`, `ShardDeleteRequestItem` is used only in `ShardDMLExecutor`. I added an assertion.

seut commented on server/src/main/java/io/crate/execution/engine/indexing/IndexWriterProjector.java:
```diff
@@ -70,6 +71,7 @@ public class IndexWriterProjector implements Projector {
     private final ShardingUpsertExecutor shardingUpsertExecutor;
 
     public IndexWriterProjector(ClusterService clusterService,
+                                DocTableInfo tableInfo,
```
Unused?

seut commented on server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java:
```diff
@@ -101,8 +102,11 @@ public class ShardingUpsertExecutor
 
     ShardingUpsertExecutor(ClusterService clusterService,
+                           @Nullable BiConsumer<String, IndexItem> constraintsChecker,
+                           @Nullable Runnable validatorsCacheCleaner,
```
I think that it's not really nice that we require here on logic which is done outside of this component. But I also have no good suggestion right now how we could add any post-completion logic outside of this component. Maybe @mfussenegger has a better idea?
As least we should change the name to be more general as for the component this is just a runnable run after completion, what it does is not known.
```suggestion
                           Runnable onCompletion,
```

BaurzhanSakhariev replied:
I think previous solution with having cache in `ShardedRequests` and **auto** clearance in `close()` had its own flaws but since a new approach not much better because of having post completion logic outside, maybe we can get back to it?

seut commented on server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java:
```diff
@@ -162,7 +170,12 @@ public TItem apply(ShardedRequests<TReq, TItem> shardedRequests, Row row, Boolea
 
             RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
             ShardLocation shardLocation = getShardLocation(indexName, id, routing);
+            assert item instanceof ShardUpsertRequest.Item : "IndexItem must be instance of ShardUpsertRequest.Item";
+            IndexItem indexItem = (IndexItem) item;
             if (shardLocation == null) {
+                // Validation is done before creating an index in order to ensure
+                // that no "bad partitions" will be left behind in case of validation failure.
+                constraintsChecker.accept(indexNameResolver.get(), indexItem);
```
Even that it currently on used by this, I don't see a logical reason to require it to be of type `IndexItem`.
```suggestion
            if (shardLocation == null) {
                // Validation is done before creating an index in order to ensure
                // that no "bad partitions" will be left behind in case of validation failure.
                if (item instanceof IndexItem indexItem) {
                    constraintsChecker.accept(indexNameResolver.get(), indexItem);
                }
```

seut commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -486,25 +504,27 @@ public List<CompletableFuture<Long>> executeBulk(DependencyCarrier dependencies,
-        var validator = validatorsCache.computeIfAbsent(
-            indexName,
-            index -> Indexer.createConstraintCheck(
+    public static BiConsumer<String, IndexItem> createConstraintsCheck(DocTableInfo tableInfo,
```
Why not leave it as a `void` return and just create a lambda on usage? 
```suggestion
    public static void checkConstraints(DocTableInfo tableInfo,
```

BaurzhanSakhariev replied:
Thx, applied it and also adjusted slightly in https://github.com/crate/crate/pull/14373/commits/bc24962547a73b116e8ffe8613d987d712e0736e

seut reviewed (approved):
👍

BaurzhanSakhariev commented:
Hi @seut, thanks for the review! 

I addressed comments but I would hold on a bit because of the cache creation/clearance approach. So far there are 2 ways but each has flaws:
1. ShardedRequests
+ auto clearance, no need to explicitly do anything, declaration/clearance happens in a single place
- usage is different place (groper)
2. Current approach
usage/declaraton/usage sort of in a single place (if we count Runnable/BiConsumer creation as a usage/declaration) but then need to propagate everywhere.

seut commented:
> I addressed comments but I would hold on a bit because of the cache creation/clearance approach. So far there are 2 ways but each has flaws:

I'd suggest to merge the now as-is. We can change the relevant code later at anytime if we come up with a better approach.
